File: sdks/csharp/src/Sudomimus.Token/IdToken.cs
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace Sudomimus.Token;

/// <summary>Header claims of an OIDC <c>id_token</c>.</summary>
/// <remarks>
/// Unlike Sudomimus access/refresh tokens, an id_token is a standard OIDC JWT:
/// <c>kid</c> identifies the platform signing key in the OIDC JWKS.
/// </remarks>
public sealed record IdTokenHeader
{
    [JsonPropertyName("alg")] public string? Algorithm { get; init; }
    [JsonPropertyName("typ")] public string? Type { get; init; }
    [JsonPropertyName("kid")] public string? KeyId { get; init; }
}

/// <summary>Body claims of a Sudomimus OIDC <c>id_token</c>.</summary>
/// <remarks>
/// Every claim lives in the JWT body (standard OIDC). <c>Subject</c> is the
/// per-(account, sector) sector subject — identical to the access-token body
/// <c>Subject</c>. The token is signed by the platform key.
/// </remarks>
public sealed record IdTokenBody
{
    [JsonPropertyName("iss")] public required string Issuer { get; init; }
    [JsonPropertyName("sub")] public required string Subject { get; init; }
    [JsonPropertyName("aud")] public required string Audience { get; init; }
    [JsonPropertyName("iat")] public required long IssuedAt { get; init; }
    [JsonPropertyName("exp")] public required long ExpiresAt { get; init; }
    [JsonPropertyName("at_hash")] public string? AtHash { get; init; }
    [JsonPropertyName("nonce")] public string? Nonce { get; init; }
    [JsonPropertyName("auth_time")] public long? AuthTime { get; init; }
    [JsonPropertyName("email")] public string? Email { get; init; }
    [JsonPropertyName("email_verified")] public bool? EmailVerified { get; init; }
    [JsonPropertyName("name")] public string? Name { get; init; }
    [JsonPropertyName("amr")] public IReadOnlyList<string>? Amr { get; init; }
    [JsonPropertyName("acr")] public string? Acr { get; init; }
}

/// <summary>Decoded response of the OIDC <c>/userinfo</c> endpoint.</summary>
public sealed record UserInfoResponse
{
    [JsonPropertyName("sub")] public required string Subject { get; init; }
    [JsonPropertyName("email")] public string? Email { get; init; }
    [JsonPropertyName("email_verified")] public bool? EmailVerified { get; init; }
    [JsonPropertyName("name")] public string? Name { get; init; }
}

/// <summary>
/// Optional expectations narrowing <see cref="IdToken.Verify"/>. A <c>null</c>
/// field is not checked; a <c>null</c> <see cref="Now"/> defaults to
/// <see cref="DateTimeOffset.UtcNow"/>.
/// </summary>
public sealed record IdTokenExpectations
{
    public string? Audience { get; init; }
    public string? Issuer { get; init; }
    public string? Nonce { get; init; }
    public DateTimeOffset? Now { get; init; }
}

/// <summary>
/// A parsed Sudomimus OIDC <c>id_token</c>. Unlike Sudomimus access/refresh
/// tokens (whose envelope claims live in the JWT header), an id_token is a
/// standard OIDC JWT: every claim lives in the body and the token is signed by
/// the <b>platform</b> key (resolve it from the OIDC JWKS by the header
/// <c>kid</c>), not by an application's signing key. <see cref="Verify"/>
/// therefore checks <c>exp</c> from the body.
/// </summary>
public sealed class IdToken
{
    private static readonly JsonSerializerOptions s_jsonOptions = new()
    {
        PropertyNameCaseInsensitive = true,
    };

    public string Raw { get; }
    public byte[] SigningInput { get; }
    public byte[] Signature { get; }
    public IdTokenHeader Header { get; }
    public IdTokenBody Body { get; }

    private IdToken(string raw, byte[] signingInput, byte[] signature, IdTokenHeader header, IdTokenBody body)
    {
        Raw = raw;
        SigningInput = signingInput;
        Signature = signature;
        Header = header;
        Body = body;
    }

    /// <summary>Verify the signature with the given PEM-encoded platform RSA public key.</summary>
    public bool VerifySignature(string publicKeyPem)
    {
        using var rsa = RSA.Create();
        rsa.ImportFromPem(publicKeyPem);
        return rsa.VerifyData(
            SigningInput,
            Signature,
            HashAlgorithmName.SHA256,
            RSASignaturePadding.Pkcs1);
    }

    /// <summary>Parse an id_token into its header and body without verifying it.</summary>
    /// <exception cref="TokenException">Thrown when the JWT is structurally invalid.</exception>
    public static IdToken Parse(string jwt)
    {
        if (string.IsNullOrEmpty(jwt))
        {
            throw new TokenException(TokenErrorCode.InvalidJwt, "Token is empty.");
        }

        var parts = jwt.Split('.');
        if (parts.Length != 3)
        {
            throw new TokenException(
                TokenErrorCode.InvalidJwt,
                $"Token must have exactly three dot-separated segments; got {parts.Length}.");
        }

        byte[] headerBytes;
        byte[] bodyBytes;
        byte[] signatureBytes;
        try
        {
            headerBytes = JwtCodec.DecodeBase64UrlSegment(parts[0]);
            bodyBytes = JwtCodec.DecodeBase64UrlSegment(parts[1]);
            signatureBytes = JwtCodec.DecodeBase64UrlSegment(parts[2]);
        }
        catch (FormatException ex)
        {
            throw new TokenException(TokenErrorCode.InvalidJwt, $"Failed to decode id_token segments: {ex.Message}");
        }

        IdTokenHeader? header;
        IdTokenBody? body;
        try
        {
            header = JsonSerializer.Deserialize<IdTokenHeader>(headerBytes, s_jsonOptions);
            body = JsonSerializer.Deserialize<IdTokenBody>(bodyBytes, s_jsonOptions);
        }
        catch (JsonException ex)
        {
            throw new TokenException(TokenErrorCode.InvalidJwt, $"Failed to deserialize id_token claims: {ex.Message}");
        }

        if (header is null || body is null)
        {
            throw new TokenException(TokenErrorCode.InvalidJwt, "id_token header or body deserialized to null.");
        }

        var signingInput = Encoding.UTF8.GetBytes($"{parts[0]}.{parts[1]}");
        return new IdToken(jwt, signingInput, signatureBytes, header, body);
    }

    /// <summary>
    /// Verify an OIDC id_token against a platform public key (resolved from the
    /// OIDC JWKS). Checks the body <c>exp</c>, the RS256 signature, and any of
    /// the supplied audience/issuer/nonce expectations.
    /// </summary>
    /// <exception cref="TokenException">Thrown on any verification failure.</exception>
    public static IdToken Verify(string jwt, string platformPublicKeyPem, IdTokenExpectations? expectations = null)
    {
        var parsed = Parse(jwt);
        var expect = expectations ?? new IdTokenExpectations();
        var now = expect.Now ?? DateTimeOffset.UtcNow;

        if (now >= DateTimeOffset.FromUnixTimeSeconds(parsed.Body.ExpiresAt))
        {
            throw new TokenException(TokenErrorCode.Expired, "id_token has expired.");
        }

        if (!parsed.VerifySignature(platformPublicKeyPem))
        {
            throw new TokenException(
                TokenErrorCode.InvalidSignature,
                "id_token signature does not match the platform public key.");
        }

        if (expect.Audience is not null && parsed.Body.Audience != expect.Audience)
        {
            throw new TokenException(TokenErrorCode.WrongAudience, "id_token aud does not match the expected client id.");
        }
        if (expect.Issuer is not null && parsed.Body.Issuer != expect.Issuer)
        {
            throw new TokenException(TokenErrorCode.WrongIssuer, "id_token iss does not match the expected issuer.");
        }
        if (expect.Nonce is not null && parsed.Body.Nonce != expect.Nonce)
        {
            throw new TokenException(TokenErrorCode.WrongNonce, "id_token nonce does not match the value sent at /authorize.");
        }

        return parsed;
    }
}

File: sdks/go/token/idtoken.go
package token

import (
	"crypto"
	"crypto/rsa"
	"crypto/sha256"
	"encoding/json"
	"strings"
	"time"
)

// IDTokenHeader holds the header claims of an OIDC id_token. Unlike Sudomimus
// access/refresh tokens, an id_token is a standard OIDC JWT: KeyID identifies
// the platform signing key in the OIDC JWKS.
type IDTokenHeader struct {
	Algorithm string `json:"alg,omitempty"`
	Type      string `json:"typ,omitempty"`
	KeyID     string `json:"kid,omitempty"`
}

// IDTokenBody holds the body claims of a Sudomimus OIDC id_token. Every claim
// lives in the JWT body (standard OIDC). Subject is the per-(account, sector)
// sector subject — identical to the access-token body Subject. The token is
// signed by the platform key, not by an application's signing key.
type IDTokenBody struct {
	Issuer        string   `json:"iss"`
	Subject       string   `json:"sub"`
	Audience      string   `json:"aud"`
	IssuedAt      int64    `json:"iat"`
	ExpiresAt     int64    `json:"exp"`
	AtHash        string   `json:"at_hash,omitempty"`
	Nonce         string   `json:"nonce,omitempty"`
	AuthTime      int64    `json:"auth_time,omitempty"`
	Email         string   `json:"email,omitempty"`
	EmailVerified bool     `json:"email_verified,omitempty"`
	Name          string   `json:"name,omitempty"`
	AMR           []string `json:"amr,omitempty"`
	ACR           string   `json:"acr,omitempty"`
}

// IDToken is a parsed OIDC id_token.
type IDToken struct {
	Raw          string
	SigningInput []byte
	Signature    []byte
	Header       IDTokenHeader
	Body         IDTokenBody
}

// UserInfo is the decoded response of the OIDC /userinfo endpoint. Subject is
// the same sector subject carried by the id_token; the other claims are
// scope-gated by the access token presented to /userinfo.
type UserInfo struct {
	Subject       string `json:"sub"`
	Email         string `json:"email,omitempty"`
	EmailVerified bool   `json:"email_verified,omitempty"`
	Name          string `json:"name,omitempty"`
}

// VerifySignature returns true when the RSA-SHA256 signature matches the given
// PEM-encoded public key.
func (t *IDToken) VerifySignature(publicKeyPEM string) bool {
	pub, err := parseRSAPublicKey(publicKeyPEM)
	if err != nil {
		return false
	}
	digest := sha256.Sum256(t.SigningInput)
	return rsa.VerifyPKCS1v15(pub, crypto.SHA256, digest[:], t.Signature) == nil
}

// ParseIDToken parses an id_token into its header and body without verifying.
func ParseIDToken(jwt string) (*IDToken, error) {
	if jwt == "" {
		return nil, newError(ErrInvalidJWT, "token is empty")
	}
	parts := strings.Split(jwt, ".")
	if len(parts) != 3 {
		return nil, newError(ErrInvalidJWT, "token must have exactly three dot-separated segments; got %d", len(parts))
	}

	headerBytes, err := b64url.DecodeString(parts[0])
	if err != nil {
		return nil, newError(ErrInvalidJWT, "failed to decode header segment: %s", err)
	}
	bodyBytes, err := b64url.DecodeString(parts[1])
	if err != nil {
		return nil, newError(ErrInvalidJWT, "failed to decode body segment: %s", err)
	}
	sigBytes, err := b64url.DecodeString(parts[2])
	if err != nil {
		return nil, newError(ErrInvalidJWT, "failed to decode signature segment: %s", err)
	}

	var header IDTokenHeader
	if err := json.Unmarshal(headerBytes, &header); err != nil {
		return nil, newError(ErrInvalidJWT, "failed to deserialize header: %s", err)
	}
	var body IDTokenBody
	if err := json.Unmarshal(bodyBytes, &body); err != nil {
		return nil, newError(ErrInvalidJWT, "failed to deserialize body: %s", err)
	}

	return &IDToken{
		Raw:          jwt,
		SigningInput: []byte(parts[0] + "." + parts[1]),
		Signature:    sigBytes,
		Header:       header,
		Body:         body,
	}, nil
}

// IDTokenExpectations narrows id_token verification. A zero-value field is not
// checked; a zero Now defaults to time.Now().
type IDTokenExpectations struct {
	Audience string
	Issuer   string
	Nonce    string
	Now      time.Time
}

// VerifyIDToken verifies an OIDC id_token against a platform public key
// (resolved from the OIDC JWKS). It checks the body exp, the RS256 signature,
// and any supplied audience/issuer/nonce expectations.
func VerifyIDToken(jwt, platformPublicKeyPEM string, expect IDTokenExpectations) (*IDToken, error) {
	parsed, err := ParseIDToken(jwt)
	if err != nil {
		return nil, err
	}

	now := expect.Now
	if now.IsZero() {
		now = time.Now()
	}
	if parsed.Body.ExpiresAt == 0 || !now.Before(time.Unix(parsed.Body.ExpiresAt, 0)) {
		return nil, newError(ErrExpired, "id_token has expired")
	}

	if !parsed.VerifySignature(platformPublicKeyPEM) {
		return nil, newError(ErrInvalidSignature, "id_token signature does not match the platform public key")
	}

	if expect.Audience != "" && parsed.Body.Audience != expect.Audience {
		return nil, newError(ErrWrongAudience, "id_token aud does not match the expected client id")
	}
	if expect.Issuer != "" && parsed.Body.Issuer != expect.Issuer {
		return nil, newError(ErrWrongIssuer, "id_token iss does not match the expected issuer")
	}
	if expect.Nonce != "" && parsed.Body.Nonce != expect.Nonce {
		return nil, newError(ErrWrongNonce, "id_token nonce does not match the value sent at /authorize")
	}

	return parsed, nil
}
